### PR TITLE
`Assessment`: Improve the manual text selection for tutors 

### PR DESCRIPTION
**Pull request by valeriebucher**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

During the assessment of text exercises, tutors can create custom text blocks by pressing `alt/option`, selecting a piece of text, and then clicking the `Assess` button. This functionality often leads to problems because the `Assess` button appears at a random position or not at all. (see screencast below)
The changes within this PR allow tutors to manually select text easier by clicking on the first and last word of the text they want to assess. 

### Description

There are multiple text block groups within a text submission; by iterating over these groups and splitting the text by spaces, we get the submission words that can be clicked on.
First, the word index within the whole submission needs to be calculated for each selected word. The selected word and its index are saved in an array. Once two words are selected, the array is emitted to the `manual-textblock-selection.component` where the corresponding text block is created.

### Steps for Testing

Prerequisites:
- 1 Tutor
- 1 Text Exercise with a submission

1. Go to the `Exercise Dashboard` of the text exercise.
2. Start the assessment on a text submission.
3. Press `alt/option` to select a custom text block.
4. Click on the first and last word of the text block you want to assess. The corresponding textblock should be created.
5. Change the order: Click the ending word first and then the starting word. The corresponding textblock should be created.
6. Check that it works if the submission contains whitespaces (linebreaks, tabs, spaces).
7. Check that nothing should happen if the tutor clicks on two words but an already created text block is inbetween (see screenshot example).

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

**Old version:**

https://user-images.githubusercontent.com/73841445/167175066-62983f94-cae5-4247-86dd-fe3029b29f92.mov

**New version:**

https://user-images.githubusercontent.com/73841445/167175100-a3042868-d62d-4def-acf0-853f2b58db0c.mov

**Example for testing step 6:**

Bildschirmfoto 2022-05-06 um 17 29 50

### CODE REVIEW

**Review by joaquinmurcia — APPROVED** [review]
Works as expected, even when in step 7 you click a word that is inside the selected block.
Good work, needed feature :smile:

**Review by bassner — CHANGES_REQUESTED** [review]
Please see my inline comment :)

**Review by alaraozdenler — DISMISSED** [review]
Tested on ts3, works as described

**Review by xlieA — DISMISSED** [review]
Tested on ts3. Everything worked as described in the testing steps. Very useful :)

File: src/main/webapp/app/exercises/text/shared/manual-text-selection/manual-text-selection.component.scss [hunk 1]
```diff
@@ -1,7 +1,20 @@
...
+span {
+    &:hover,
+    &:focus {
+        background: $gray-300;
+        padding: 2px 4px;
+        border-radius: 4px;
+        cursor: default;
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (valeriebucher) [reply src/main/webapp/app/exercises/text/shared/manual-text-selection/manual-text-selection.component.scss]: Thank you for the detailed description! I adapted it.

**Review by nadinesamyy — APPROVED** [review]
Tested on TS3, all reported enhancements work as expected

**Review by nadinesamyy — DISMISSED** [review]

**Review by bassner — APPROVED** [review]
Thanks for the updates! Code looks good to me now.
Tested the feature on TS5, works as intended.
Good job 👍

**Review by ge65cer — DISMISSED** [review]
Everything works alright. I get a bit of a weird behavior when the submission has multiple subsequent spaces, but nothing breaks.

Screenshot

**Review by joaquinmurcia — APPROVED** [review]

**Review by Santia-go — DISMISSED** [review]
Tested on TS1
Working as expected!
The corresponding textblock is created even with whitespaces.

**Review by jschmautzer — APPROVED** [review]
tested on ts1, worked as expected, great feature!

**Review by jbllhff — APPROVED** [review]
tested on ts2, thanks!

**Review by mallgaier — DISMISSED** [review]
Tested, genuinely nice enhancement
Excellent work 👍

**Review by max-bergmann — APPROVED** [review]
Tested on TS3. Works as expected

File: src/main/webapp/i18n/de/textAssessment.json [hunk 2]
```diff
@@ -59,7 +59,7 @@
                     "reused": "{{reusedCount}} mal wiederverwendet"
                 }
             },
-            "assessmentInstruction": "Bitte bewege die Maus über den zu bewertenden Textblock und klicke diesen an, um ihn zu bewerten.<br>Drücke <code>alt</code>/<code>option</code> und wähle gleichzeitig Text aus, um eigene Textblöcke zu bewerten.",
+            "assessmentInstruction": "Bitte bewege die Maus über den zu bewertenden Textblock und klicke diesen an, um ihn zu bewerten.<br>Um eigene Textblöcke zu bewerten, drücke <code>alt</code>/<code>option</code> und klicke in beliebiger Reihenfolge auf das erste und letzte Wort eines Textabchnitts.",
```
Comment (maximiliansoelch) [inline src/main/webapp/i18n/de/textAssessment.json:62]:
```suggestion
            "assessmentInstruction": "Bitte bewege die Maus über den zu bewertenden Textblock und klicke diesen an, um ihn zu bewerten.<br>Um eigene Textblöcke zu bewerten, drücke <code>alt</code>/<code>option</code> und klicke in beliebiger Reihenfolge auf das erste und letzte Wort eines Textabschnitts.",
```

**Review by maximiliansoelch — CHANGES_REQUESTED** [review]
Tested on TS1, feature works as expected 👍
But there is a small typo in your German translation string.

**Review by maximiliansoelch — APPROVED** [review]

**Review by ge65cer — APPROVED** [review]

**Review by mallgaier — APPROVED** [review]
Re-approve after typo fixing

**Review by rhubinger — APPROVED** [review]
Tested on TS2. Worked as described.

**Review by bassner — APPROVED** [review]
Reapprove (typo)

**Review by Santia-go — APPROVED** [review]
Tested on TS1
Working as expected!
The corresponding text block is created even with whitespaces.

**Pull request merged**